lib/migrations/migrations.js

```diff
@@ -0,0 +1,350 @@
+        state.lastMigration + 1);
+    }
+
+    let hasMigrated = false;
+    for (let i = state.lastMigration + 1; i <= lastID; i++) {
+      const currentMigration = new this.migrations[i](this.options);
+      const type = await currentMigration.check();
+
+      switch (type)  {
+        case types.FAKE_MIGRATE: {
+          this.logger.info('Does not apply, fake migrating %d.', i);
+          state.lastMigration = i;
+          await this.saveState(state);
+          break;
+        }
+        case types.SKIP: {
+          this.logger.info('Can not migrate, skipping %d.', i);
+          this.logger.warning(currentMigration.warning());
+          state.lastMigration = i;
+          state.skipped.push(i);
+          await this.saveState(state);
+          break;
+        }
+        case types.MIGRATE: {
+          if (!this.options.migrateFlag)
+            throw new Error('Database needs migration.');
+
+          state.inProgress = true;
+          await this.saveState(state);
+
+          this.logger.info('Migration %d in progress..', i);
+          const batch = this.ldb.batch();
+          await currentMigration.migrate(batch);
+
+          state.inProgress = false;
+          state.lastMigration = i;
+          this.writeState(batch, state);
+
+          await batch.write();
+          this.logger.info('Migration %d is done.', i);
+          hasMigrated = true;
+          break;
+        }
+        default:
+          throw new Error('Unknown migration type.');
+      }
+    }
+
+    if (!hasMigrated && this.options.migrateFlag)
+      throw new Error('There are no available migrations.');
+  }
+
+  /**
+   * Ensure we have migration entry in DB.
+   * Do migration of the old migration states.
+   */
+
+  async ensure() {
+    const migrations = await this.ldb.get(this.layout.M.encode());
+
+    // This could mean two things:
+    //  1. database is new.
+    //  2. we are migrating from old version.
+    if (!migrations) {
+      const b = this.ldb.batch();
+      const state = new MigrationState();
+
+      // move migration entries if we are migrating
+      // from old version with Migrations.
+      // p.s. if there are no migrations:
+      //  1. database is new
+      //  2. we are migrating from even older version.
+      const oldMigrations = await this.ldb.keys({
+        gte: oldLayout.M.min(),
+        lte: oldLayout.M.max(),
+        parse: key => oldLayout.M.decode(key)[0]
+      });
+
+      let max = -1;
+      for (const id of oldMigrations) {
+        b.del(oldLayout.M.encode(id));
+        max = Math.max(id, max);
+      }
+
+      if (max > -1)
+        state.lastMigration = max;
+
+      b.put(this.layout.M.encode(), state.encode());
+      await b.write();
+      return;
+    }
+  }
+
+  /**
+   * Get max migration ID from the map
+   * @returns {Number}
+   */
+
+  getLastMigrationID() {
+    const ids = Object.keys(this.migrations);
+
+    if (ids.length === 0)
+      return 0;
```

Comment:
Same here -- since we used migration ID `0` for  https://github.com/handshake-org/hsd/pull/396, maybe "no migrations" should be represented by `-1`.

Reply:
ChainDB has merged 0 and 1 migrations together, so even if you have migration 0 in there it does not matter you still need to migrate. hm.. 
But walletdb currently uses 0 as "lastMigration".
We will need to change from varint to int32.

Will look into this.